File: backend/src/main/java/com/evil/backend/user/entity/Account.java
package com.evil.backend.user.entity;
import java.time.LocalDateTime;
import java.util.Set;

import com.fasterxml.jackson.annotation.JsonIgnore;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
public class Account {
    @Id
    @GeneratedValue
    private Long id;
    // user information
    @Column(nullable = false, unique = true, length = 100)
    private String username;
    @Column(nullable = false, unique = true)
    private String email;
    @Column(nullable = false)
    private String firstName;
    @Column(nullable = false)
    private String lastName;
    @Column(nullable = false)
    private String phoneNumber;
    @JsonIgnore
    @Column(nullable = false)

    // system access
    private String password;
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
        name = "account_roles",
        joinColumns = @JoinColumn(name = "account_id"),
        inverseJoinColumns = @JoinColumn(name = "role_id")
    )
    private boolean validated;
    private boolean locked;
    private Set<Role> roles;

    // handy information
    @Column(length = 100)
    @ManyToOne
    private JobTitle jobTitle;
    private LocalDateTime accountCreated;
    private LocalDateTime lastLogin;
}

File: frontend/shared-assets/interfaces/Account.ts
// PrivilegeType enum
export enum PrivilegeType {
  READ_ACCESS = "READ_ACCESS",
  MANAGE_SERVER = "MANAGE_SERVER",
  WRITE_ACCESS = "WRITE_ACCESS",
  KICK_USERS = "KICK_USERS",
  BAN_USERS = "BAN_USERS",
  VIEW_LOGS = "VIEW_LOGS",
  EDIT_PROFILE = "EDIT_PROFILE"
}

// Privilege interface
export interface Privilege {
  id: number;
  displayName: string;
  type: PrivilegeType;
  description: string;
}

// Role interface
export interface Role {
  id: number;
  name: string;
  color: string | null;
  privileges: Privilege[];
}

// JobTitle interface
export interface JobTitle {
  id: number;
  title: string;
  emote: string;
}

// AccountDto interface
export interface AccountDto {
  id: number;
  username: string;
  email: string;
  firstName: string;
  lastName: string;
  phoneNumber: string;
  roles: Role[];
  validated: boolean;
  locked: boolean;
  jobTitle: JobTitle | null;
  accountCreated: string;
  lastLogin: string | null;
}
